plugins/Live/VisitorDetails.php

```diff
@@ -181,12 +183,42 @@ public function finalizeProfile($visits, &$profile)
         $profile['totalVisitDurationPretty'] = $formatter->getPrettyTimeFromSeconds($profile['totalVisitDuration'], true);
 
         $rows                        = $visits->getRows();
+
+        $firstVisit = $profile['visit_first'];
+        if (count($rows) >= Config::getInstance()->General['live_visitor_profile_max_visits_to_aggregate']) {
+            $firstVisit = $this->fetchFirstVisit();
+        }
+
         $profile['userId']           = $visits->getLastRow()->getColumn('userId');
-        $profile['firstVisit']       = $this->getVisitorProfileVisitSummary(end($rows));
-        $profile['lastVisit']        = $this->getVisitorProfileVisitSummary(reset($rows));
+        $profile['firstVisit']       = $this->getVisitorProfileVisitSummary($firstVisit);
+        $profile['lastVisit']        = $this->getVisitorProfileVisitSummary($profile['visit_last']);
         $profile['visitsAggregated'] = count($rows);
     }
 
+    /**
+     * Fetch first visit from Live API
+     *
+     * @return DataTable\Row
+     */
+    private function fetchFirstVisit()
+    {
+        $response = Request::processRequest('Live.getFirstVisitForVisitorId', [
+            'idSite' => $this->getIdSite(),
+            'visitorId' => $this->getVisitorId(),
+        ]);
+
+        $row                                 = $response->getFirstRow();
+        $visitorData                         = $row->getColumns();
+        $visitorData['firstActionTimestamp'] = strtotime($visitorData['visit_first_action_time']);
+
+        // process referrer details
+        $referrersDetails = new \Piwik\Plugins\Referrers\VisitorDetails();
+        $referrersDetails->setDetails($visitorData);
+        $referrersDetails->extendVisitorDetails($visitorData);
```

Comment:
this might be a bit hackish, but the referrer information is displayed for that "last visit". As it is fetched afterwards it does not go through the visitordetails process and such leaks the processed information.
If we later decide to display some other information processed by another plugin we might need to do that with additional or even all `VisitorDetails` classes